# Relational Databases



In [1]:
pip install sqlalchemy psycopg2 ipython-sql

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 384.9/384.9 kB 7.1 MB/s eta 0:00:00a 0:00:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 29.8 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 660.5/660.5 kB 32.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 14.2 MB/s eta 0:00:00
  Created wheel for psycopg2: filename=psycopg2-2.9.9-cp312-cp312-linux_aarch64.whl size=510611 sha256=bdaa36de0c02ba11af7c376f30d52da8b61afb1ef4107bf86331eaca476989e0
  Stored in directory: /home/vscode/.cache/pip/wheels/ff/ac/80/7ccec163e3d05ae2112311b895132409b9abfd7e0c1c6b5183
Successfully built psycopg2
Note: you may need to restart the kernel to use updated packages.


In [2]:
%load_ext sql


In [3]:
%sql postgresql://postgres:postgres@localhost/local


Traceback (most recent call last):
  File "/workspaces/introduction-to-sql/.venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 146, in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/introduction-to-sql/.venv/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 3304, in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
  File "/workspaces/introduction-to-sql/.venv/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 449, in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/introduction-to-sql/.venv/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 1263, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/introduction-to-sql/.venv/lib/python3.12/site-packages/sqlalchemy/pool/base.py", line 712, in c

In [4]:
%config SqlMagic.autolimit = 10


## Our very own table

A database has been set up for this course and the `books` table is
available here.

Run the code to explore what data `books` holds!

**Instructions**

- Hit "Run Code" to see the `books` table.

**Answer**


In [10]:
%%sql
SELECT *
FROM books;


 * postgresql://postgres:***@localhost/local
350 rows affected.


[(1, '10-Day Green Smoothie Cleanse', 'JJ Smith', 2016, 'Non Fiction'),
 (2, '11/22/63: A Novel', 'Stephen King', 2011, 'Fiction'),
 (3, '12 Rules for Life: An Antidote to Chaos', 'Jordan B. Peterson', 2018, 'Non Fiction'),
 (4, '1984 (Signet Classics)', 'George Orwell', 2017, 'Fiction'),
 (5, '"5,000 Awesome Facts (About Everything!) (National Geographic Kids)"', 'National Geographic Kids', 2019, 'Non Fiction'),
 (6, 'A Dance with Dragons (A Song of Ice and Fire)', 'George R. R. Martin', 2011, 'Fiction'),
 (7, 'A Game of Thrones / A Clash of Kings / A Storm of Swords / A Feast of Crows / A Dance with Dragons', 'George R. R. Martin', 2014, 'Fiction'),
 (8, 'A Gentleman in Moscow: A Novel', 'Amor Towles', 2017, 'Fiction'),
 (9, '"A Higher Loyalty: Truth, Lies, and Leadership"', 'James Comey', 2018, 'Non Fiction'),
 (10, 'A Man Called Ove: A Novel', 'Fredrik Backman', 2016, 'Fiction'),
 (11, "A Patriot's History of the United States: From Columbus's Great Discovery to the War on Terror", 'Larry Schweikart', 2010, 'Non Fiction'),
 (12, 'A Stolen Life: A Memoir', 'Jaycee Dugard', 2011, 'Non Fiction'),
 (13, 'A Wrinkle in Time (Time Quintet)', "Madeleine L'Engle", 2018, 'Fiction'),
 (14, '"Act Like a Lady, Think Like a Man: What Men Really Think About Love, Relationships, Intimacy, and Commitment"', 'Steve Harvey', 2009, 'Non Fiction'),
 (15, '"Adult Coloring Book Designs: Stress Relief Coloring Book: Garden Designs, Mandalas, Animals, and Paisley Patterns"', 'Adult Coloring Book Designs', 2016, 'Non Fiction'),
 (16, 'Adult Coloring Book: Stress Relieving Animal Designs', 'Blue Star Coloring', 2015, 'Non Fiction'),
 (17, 'Adult Coloring Book: Stress Relieving Patterns', 'Blue Star Coloring', 2015, 'Non Fiction'),
 (18, '"Adult Coloring Books: A Coloring Book for Adults Featuring Mandalas and Henna Inspired Flowers, Animals, and Paisley…"', 'Coloring Books for Adults', 2015, 'Non Fiction'),
 (19, 'Alexander Hamilton', 'Ron Chernow', 2016, 'Non Fiction'),
 (20, 'All the Light We Cannot See', 'Anthony Doerr', 2014, 'Fiction'),
 (21, 'Allegiant', 'Veronica Roth', 2013, 'Fiction'),
 (22, 'American Sniper: The Autobiography of the Most Lethal Sniper in U.S. Military History', 'Chris Kyle', 2015, 'Non Fiction'),
 (23, 'And the Mountains Echoed', 'Khaled Hosseini', 2013, 'Fiction'),
 (24, 'Arguing with Idiots: How to Stop Small Minds and Big Government', 'Glenn Beck', 2009, 'Non Fiction'),
 (25, 'Astrophysics for People in a Hurry', 'Neil deGrasse Tyson', 2017, 'Non Fiction'),
 (26, '"Autobiography of Mark Twain, Vol. 1"', 'Mark Twain', 2010, 'Non Fiction'),
 (27, 'Baby Touch and Feel: Animals', 'DK', 2015, 'Non Fiction'),
 (28, "Balance (Angie's Extreme Stress Menders)", 'Angie Grace', 2015, 'Non Fiction'),
 (29, 'Barefoot Contessa Foolproof: Recipes You Can Trust: A Cookbook', 'Ina Garten', 2012, 'Non Fiction'),
 (30, '"Barefoot Contessa, How Easy Is That?: Fabulous Recipes & Easy Tips"', 'Ina Garten', 2010, 'Non Fiction'),
 (31, 'Becoming', 'Michelle Obama', 2018, 'Non Fiction'),
 (32, 'Being Mortal: Medicine and What Matters in the End', 'Atul Gawande', 2015, 'Non Fiction'),
 (33, 'Between the World and Me', 'Ta-Nehisi Coates', 2015, 'Non Fiction'),
 (34, 'Born to Run', 'Bruce Springsteen', 2016, 'Non Fiction'),
 (35, '"Breaking Dawn (The Twilight Saga, Book 4)"', 'Stephenie Meyer', 2009, 'Fiction'),
 (36, '"Broke: The Plan to Restore Our Trust, Truth and Treasure"', 'Glenn Beck', 2010, 'Non Fiction'),
 (37, '"Brown Bear, Brown Bear, What Do You See?"', 'Bill Martin Jr.', 2017, 'Fiction'),
 (38, '"Cabin Fever (Diary of a Wimpy Kid, Book 6)"', 'Jeff Kinney', 2011, 'Fiction'),
 (39, 'Calm the F*ck Down: An Irreverent Adult Coloring Book (Irreverent Book Series)', "Sasha O'Hara", 2016, 'Non Fiction'),
 (40, "Can't Hurt Me: Master Your Mind and Defy the Odds", 'David Goggins', 2019, 'Non Fiction'),
 (41, 'Capital in the Twenty First Century', 'Thomas Piketty', 2014, 'Non Fiction'),
 (42, 'Catching Fire (The Hunger Games)', 'Suzanne Collins

# Querying

## Querying the books table

You're ready to practice writing your first SQL queries using the
`SELECT` and `FROM` keywords. Recall from the video that `SELECT` is
used to choose the fields that will be included in the result set, while
`FROM` is used to pick the table in which the fields are listed.

Feel free to explore `books` in the exercise.

Your task in this exercise is to practice selecting fields from `books`.

**Instructions**

- Use SQL to return a result set of all book titles included in the
  `books` table.

<!-- -->

- Select both the `title` and `author` fields from `books`.

<!-- -->

- Select all fields from the `books` table.

**Answer**


In [5]:
%%sql
-- Return all titles from the books table
SELECT title
FROM books;

-- Select title and author from the books table
SELECT title, author
FROM books;

-- Select all fields from the books table
SELECT *
FROM books;


 * postgresql://postgres:***@localhost/local
350 rows affected.
350 rows affected.
350 rows affected.


[(1, '10-Day Green Smoothie Cleanse', 'JJ Smith', 2016, 'Non Fiction'),
 (2, '11/22/63: A Novel', 'Stephen King', 2011, 'Fiction'),
 (3, '12 Rules for Life: An Antidote to Chaos', 'Jordan B. Peterson', 2018, 'Non Fiction'),
 (4, '1984 (Signet Classics)', 'George Orwell', 2017, 'Fiction'),
 (5, '"5,000 Awesome Facts (About Everything!) (National Geographic Kids)"', 'National Geographic Kids', 2019, 'Non Fiction'),
 (6, 'A Dance with Dragons (A Song of Ice and Fire)', 'George R. R. Martin', 2011, 'Fiction'),
 (7, 'A Game of Thrones / A Clash of Kings / A Storm of Swords / A Feast of Crows / A Dance with Dragons', 'George R. R. Martin', 2014, 'Fiction'),
 (8, 'A Gentleman in Moscow: A Novel', 'Amor Towles', 2017, 'Fiction'),
 (9, '"A Higher Loyalty: Truth, Lies, and Leadership"', 'James Comey', 2018, 'Non Fiction'),
 (10, 'A Man Called Ove: A Novel', 'Fredrik Backman', 2016, 'Fiction'),
 (11, "A Patriot's History of the United States: From Columbus's Great Discovery to the War on Terror", 'Larry Schweikart', 2010, 'Non Fiction'),
 (12, 'A Stolen Life: A Memoir', 'Jaycee Dugard', 2011, 'Non Fiction'),
 (13, 'A Wrinkle in Time (Time Quintet)', "Madeleine L'Engle", 2018, 'Fiction'),
 (14, '"Act Like a Lady, Think Like a Man: What Men Really Think About Love, Relationships, Intimacy, and Commitment"', 'Steve Harvey', 2009, 'Non Fiction'),
 (15, '"Adult Coloring Book Designs: Stress Relief Coloring Book: Garden Designs, Mandalas, Animals, and Paisley Patterns"', 'Adult Coloring Book Designs', 2016, 'Non Fiction'),
 (16, 'Adult Coloring Book: Stress Relieving Animal Designs', 'Blue Star Coloring', 2015, 'Non Fiction'),
 (17, 'Adult Coloring Book: Stress Relieving Patterns', 'Blue Star Coloring', 2015, 'Non Fiction'),
 (18, '"Adult Coloring Books: A Coloring Book for Adults Featuring Mandalas and Henna Inspired Flowers, Animals, and Paisley…"', 'Coloring Books for Adults', 2015, 'Non Fiction'),
 (19, 'Alexander Hamilton', 'Ron Chernow', 2016, 'Non Fiction'),
 (20, 'All the Light We Cannot See', 'Anthony Doerr', 2014, 'Fiction'),
 (21, 'Allegiant', 'Veronica Roth', 2013, 'Fiction'),
 (22, 'American Sniper: The Autobiography of the Most Lethal Sniper in U.S. Military History', 'Chris Kyle', 2015, 'Non Fiction'),
 (23, 'And the Mountains Echoed', 'Khaled Hosseini', 2013, 'Fiction'),
 (24, 'Arguing with Idiots: How to Stop Small Minds and Big Government', 'Glenn Beck', 2009, 'Non Fiction'),
 (25, 'Astrophysics for People in a Hurry', 'Neil deGrasse Tyson', 2017, 'Non Fiction'),
 (26, '"Autobiography of Mark Twain, Vol. 1"', 'Mark Twain', 2010, 'Non Fiction'),
 (27, 'Baby Touch and Feel: Animals', 'DK', 2015, 'Non Fiction'),
 (28, "Balance (Angie's Extreme Stress Menders)", 'Angie Grace', 2015, 'Non Fiction'),
 (29, 'Barefoot Contessa Foolproof: Recipes You Can Trust: A Cookbook', 'Ina Garten', 2012, 'Non Fiction'),
 (30, '"Barefoot Contessa, How Easy Is That?: Fabulous Recipes & Easy Tips"', 'Ina Garten', 2010, 'Non Fiction'),
 (31, 'Becoming', 'Michelle Obama', 2018, 'Non Fiction'),
 (32, 'Being Mortal: Medicine and What Matters in the End', 'Atul Gawande', 2015, 'Non Fiction'),
 (33, 'Between the World and Me', 'Ta-Nehisi Coates', 2015, 'Non Fiction'),
 (34, 'Born to Run', 'Bruce Springsteen', 2016, 'Non Fiction'),
 (35, '"Breaking Dawn (The Twilight Saga, Book 4)"', 'Stephenie Meyer', 2009, 'Fiction'),
 (36, '"Broke: The Plan to Restore Our Trust, Truth and Treasure"', 'Glenn Beck', 2010, 'Non Fiction'),
 (37, '"Brown Bear, Brown Bear, What Do You See?"', 'Bill Martin Jr.', 2017, 'Fiction'),
 (38, '"Cabin Fever (Diary of a Wimpy Kid, Book 6)"', 'Jeff Kinney', 2011, 'Fiction'),
 (39, 'Calm the F*ck Down: An Irreverent Adult Coloring Book (Irreverent Book Series)', "Sasha O'Hara", 2016, 'Non Fiction'),
 (40, "Can't Hurt Me: Master Your Mind and Defy the Odds", 'David Goggins', 2019, 'Non Fiction'),
 (41, 'Capital in the Twenty First Century', 'Thomas Piketty', 2014, 'Non Fiction'),
 (42, 'Catching Fire (The Hunger Games)', 'Suzanne Collins

## Making queries DISTINCT

You've learned that the `DISTINCT` keyword can be used to return unique
values in a field. In this exercise, you'll use this understanding to
find out more about the `books` table!

There are 350 books in the `books` table, representing all of the books
that our local library has available for checkout. But how many
different authors are represented in these 350 books? The answer is
surely less than 350. For example, J.K. Rowling wrote all seven Harry
Potter books, so if our library has all Harry Potter books, seven books
will be written by J.K Rowling. There are likely many more repeat
authors!

**Instructions**

- Write SQL code that returns a result set with just one column listing
  the unique authors in the `books` table.

<!-- -->

- Update the code to return the unique `author` and `genre` combinations
  in the `books` table.

**Answer**


In [6]:
%%sql
-- Select unique authors from the books table
SELECT DISTINCT author
FROM books;

-- Select unique authors and genre combinations from the books table
SELECT DISTINCT author, genre
FROM books;


 * postgresql://postgres:***@localhost/local
247 rows affected.
249 rows affected.


[('Eben Alexander', 'Non Fiction'),
 ('Adam Mansbach', 'Fiction'),
 ('Garth Stein', 'Fiction'),
 ('Gillian Flynn', 'Fiction'),
 ('Tucker Carlson', 'Non Fiction'),
 ('Atul Gawande', 'Non Fiction'),
 ('Wizards RPG Team', 'Fiction'),
 ('Giles Andreae', 'Fiction'),
 ('Dav Pilkey', 'Fiction'),
 ('Malcolm Gladwell', 'Non Fiction'),
 ('Craig Smith', 'Fiction'),
 ('Angie Grace', 'Non Fiction'),
 ('Rebecca Skloot', 'Non Fiction'),
 ('David Goggins', 'Non Fiction'),
 ('Patrick Thorpe', 'Fiction'),
 ('Bob Woodward', 'Non Fiction'),
 ('Tara Westover', 'Non Fiction'),
 ('Randall Munroe', 'Non Fiction'),
 ('Sara Gruen', 'Fiction'),
 ('Ta-Nehisi Coates', 'Non Fiction'),
 ('Dan Brown', 'Fiction'),
 ('Randy Pausch', 'Non Fiction'),
 ('David McCullough', 'Non Fiction'),
 ('Bill Simmons', 'Non Fiction'),
 ('Jill Twiss', 'Fiction'),
 ('Paula McLain', 'Fiction'),
 ('Patrick Lencioni', 'Non Fiction'),
 ('Heidi Murkoff', 'Non Fiction'),
 ('Mary Ann Shaffer', 'Fiction'),
 ('Scholastic', 'Non Fiction'),
 ("Bill O'Reilly", 'Non Fiction'),
 ('Jay Asher', 'Fiction'),
 ('George W. Bush', 'Non Fiction'),
 ('DK', 'Non Fiction'),
 ('Daniel H. Pink', 'Non Fiction'),
 ('Keith Richards', 'Non Fiction'),
 ('Ray Bradbury', 'Fiction'),
 ('Sarah Palin', 'Non Fiction'),
 ('Thomas Campbell', 'Non Fiction'),
 ('Bessel van der Kolk M.D.', 'Non Fiction'),
 ('Jim Collins', 'Non Fiction'),
 ('George R.R. Martin', 'Fiction'),
 ('Eric Larson', 'Non Fiction'),
 ('Mike Moreno', 'Non Fiction'),
 ('Robert Jordan', 'Fiction'),
 ('Hillary Rodham Clinton', 'Non Fiction'),
 ('Chris Cleave', 'Fiction'),
 ('Andy Weir', 'Fiction'),
 ('Emily Winfield Martin', 'Fiction'),
 ('Ree Drummond', 'Non Fiction'),
 ('Lysa TerKeurst', 'Non Fiction'),
 ('Dr. Seuss', 'Fiction'),
 ('Marty Noble', 'Non Fiction'),
 ('Mark Owen', 'Non Fiction'),
 ('Fredrik Backman', 'Fiction'),
 ('Charles Krauthammer', 'Non Fiction'),
 ('JJ Smith', 'Non Fiction'),
 ('Kathryn Stockett', 'Fiction'),
 ('Gary Chapman', 'Non Fiction'),
 ('John Grisham', 'Fiction'),
 ('David Platt', 'Non Fiction'),
 ('Walter Isaacson', 'Non Fiction'),
 ('Mark Twain', 'Non Fiction'),
 ('Gallup', 'Non Fiction'),
 ('Margaret Wise Brown', 'Fiction'),
 ('Greg Mortenson', 'Non Fiction'),
 ('The College Board', 'Non Fiction'),
 ('RH Disney', 'Fiction'),
 ('Rush Limbaugh', 'Fiction'),
 ('Laura Hillenbrand', 'Non Fiction'),
 ('Ian K. Smith M.D.', 'Non Fiction'),
 ('W. Cleon Skousen', 'Non Fiction'),
 ('Samin Nosrat', 'Non Fiction'),
 ('George R. R. Martin', 'Fiction'),
 ('Carol S. Dweck', 'Non Fiction'),
 ('F. A. Hayek', 'Non Fiction'),
 ('Chris Kyle', 'Non Fiction'),
 ('Marjorie Sarnat', 'Non Fiction'),
 ('Zhi Gang Sha', 'Non Fiction'),
 ('Timothy Ferriss', 'Non Fiction'),
 ('Coloring Books for Adults', 'Non Fiction'),
 ('The Staff of The Late Show with', ''),
 ('M Prefontaine', 'Non Fiction'),
 ('Mark R. Levin', 'Non Fiction'),
 ('Bruce Springsteen', 'Non Fiction'),
 ('Edward M. Kennedy', 'Non Fiction'),
 ('Marie Kondō', 'Non Fiction'),
 ('Gayle Forman', 'Fiction'),
 ('Suzanne Collins', 'Fiction'),
 ('J.K. Rowling', 'Fiction'),
 ('Jeff Kinney', 'Fiction'),
 ('Sherri Duskey Rinker', 'Fiction'),
 ('Joanna Gaines', 'Non Fiction'),
 ('Silly Bear', 'Non Fiction'),
 ('Harper Lee', 'Fiction'),
 ('Michael Lewis', 'Non Fiction'),
 ('Alex Michaelides', 'Fiction'),
 ('Rupi Kaur', 'Non Fiction'),
 ('Melissa Hartwig Urban', 'Non Fiction'),
 ('National Geographic Kids', 'Non Fiction'),
 ('Margaret Atwood', 'Fiction'),
 ('Golden Books', 'Fiction'),
 ('Dale Carnegie', 'Non Fiction'),
 ('Muriel Barbery', 'Fiction'),
 ('Doug Lemov', 'Non Fiction'),
 ('Pierre Dukan', 'Non Fiction'),
 ('Ken Follett', 'Fiction'),
 ('The Washington Post', 'Non Fiction'),
 ('John Heilemann', 'Non Fiction'),
 ('J. K. Rowling', 'Fiction'),
 ('Jennifer Smith', 'Non Fiction'),
 ('Paper Peony Press', 'Non Fiction'),
 ('David Zinczenko', 'Non Fiction'),
 ('Anthony Doerr', 'Fiction'),
 ('Bill Martin Jr.', 'Fiction'),
 ('Michelle Obama', 'Non Fiction'),
 ('Howard Stern', 'Non Fiction'),
 ('Jon S

## Aliasing

While the default column names in a SQL result set come from the fields
they are created from, you've learned that aliasing can be used to
rename these result set columns. This can be helpful for clarifying the
intent or contents of the column.

Your task in this exercise is to incorporate an alias into one of the
SQL queries that you worked with in the previous exercise!

**Instructions**

- Add an alias to the SQL query to rename the `author` column to
  `unique_author` in the result set.

**Answer**


In [7]:
%%sql
-- Alias author so that it becomes unique_author
SELECT DISTINCT author AS unique_author
FROM books;


 * postgresql://postgres:***@localhost/local
247 rows affected.


[('John Heilemann',),
 ('Sheryl Sandberg',),
 ('Brené Brown',),
 ('DK',),
 ('Rebecca Skloot',),
 ('Crispin Boyer',),
 ('Daniel James Brown',),
 ('Bruce Springsteen',),
 ('Fredrik Backman',),
 ('Heidi Murkoff',),
 ('Dan Brown',),
 ('Daniel H. Pink',),
 ('Jay Asher',),
 ('Jordan B. Peterson',),
 ('Rod Campbell',),
 ('William Davis',),
 ('Alex Michaelides',),
 ('Scholastic',),
 ('Ina Garten',),
 ('Jennifer Smith',),
 ('Nathan W. Pyle',),
 ('School Zone',),
 ('Chip Gaines',),
 ('Francis Chan',),
 ('Tara Westover',),
 ('Hillary Rodham Clinton',),
 ('Greg Mortenson',),
 ('Thomas Piketty',),
 ('Melissa Hartwig Urban',),
 ('Delia Owens',),
 ('David Goggins',),
 ('Don Miguel Ruiz',),
 ('Michael Lewis',),
 ('J. D. Vance',),
 ('Adam Mansbach',),
 ('Randall Munroe',),
 ('David Platt',),
 ('Edward Klein',),
 ('Gary Chapman',),
 ('Mark Manson',),
 ('Anthony Bourdain',),
 ('M Prefontaine',),
 ('Jonathan Cahn',),
 ('Joel Fuhrman MD',),
 ('George R.R. Martin',),
 ('Naomi Kleinberg',),
 ('Gillian Flynn',),
 ('Tucker Carlson',),
 ('Pretty Simple Press',),
 ('Abraham Verghese',),
 ('Neil deGrasse Tyson',),
 ('Patrick Thorpe',),
 ('Eben Alexander',),
 ('Emily Winfield Martin',),
 ('Kathryn Stockett',),
 ('Giles Andreae',),
 ('Sarah Young',),
 ('Ken Follett',),
 ('Khaled Hosseini',),
 ('Gallup',),
 ('Chris Cleave',),
 ('Randy Pausch',),
 ('Lysa TerKeurst',),
 ('Paulo Coelho',),
 ('Eric Carle',),
 ('Jeff Kinney',),
 ('Mary Ann Shaffer',),
 ('Ernest Cline',),
 ('Rick Riordan',),
 ('Ta-Nehisi Coates',),
 ('Atul Gawande',),
 ('Edward M. Kennedy',),
 ('Rob Elliott',),
 ('Raina Telgemeier',),
 ('Tony Hsieh',),
 ('J.K. Rowling',),
 ('Geneen Roth',),
 ('Jon Stewart',),
 ('Adult Coloring Book Designs',),
 ('RH Disney',),
 ('Carol S. Dweck',),
 ('Celeste Ng',),
 ('Admiral William H. McRaven',),
 ('William P. Young',),
 ('Hopscotch Girls',),
 ('Tatiana de Rosnay',),
 ('Pete Souza',),
 ('Audrey Niffenegger',),
 ('Jim Collins',),
 ('Charlaine Harris',),
 ('Delegates of the Constitutional',),
 ('Julia Child',),
 ('Marie Kondō',),
 ('Brandon Stanton',),
 ('Dr. Steven R Gundry MD',),
 ('Harper Lee',),
 ('Keith Richards',),
 ('Johanna Basford',),
 ('Andy Weir',),
 ('Amor Towles',),
 ('Allie Brosh',),
 ('Paula Hawkins',),
 ('Blue Star Coloring',),
 ('Patrick Lencioni',),
 ('Stieg Larsson',),
 ('Anthony Doerr',),
 ('Chris Kyle',),
 ('Paul Kalanithi',),
 ('Dr. Seuss',),
 ('Kevin Kwan',),
 ('Justin Halpern',),
 ('W. Cleon Skousen',),
 ('Maurice Sendak',),
 ('Laura Hillenbrand',),
 ('Glenn Beck',),
 ('Angie Grace',),
 ('Charles Duhigg',),
 ('Deborah Diesen',),
 ('Dave Ramsey',),
 ('Jill Twiss',),
 ('Pierre Dukan',),
 ('Timothy Ferriss',),
 ('Rush Limbaugh',),
 ('Ree Drummond',),
 ('Elie Wiesel',),
 ('Ray Bradbury',),
 ('Sandra Boynton',),
 ('Coloring Books for Adults',),
 ('Doug Lemov',),
 ('J. K. Rowling',),
 ('Roger Priddy',),
 ('The Washington Post',),
 ("Bill O'Reilly",),
 ('Michelle Obama',),
 ('Mark Hyman M.D.',),
 ('Rob Bell',),
 ('Rupi Kaur',),
 ('Elizabeth Strout',),
 ('Dale Carnegie',),
 ('Dav Pilkey',),
 ('National Geographic Kids',),
 ('Daniel Lipkowitz',),
 ('Robert Munsch',),
 ('Bob Woodward',),
 ('George R. R. Martin',),
 ('Mark R. Levin',),
 ('JJ Smith',),
 ('Golden Books',),
 ('Bessel van der Kolk M.D.',),
 ('Adam Gasiewski',),
 ("Sasha O'Hara",),
 ('Larry Schweikart',),
 ('George W. Bush',),
 ('Lin-Manuel Miranda',),
 ('Samin Nosrat',),
 ('Marjorie Sarnat',),
 ('Garth Stein',),
 ('George Orwell',),
 ('Susan Cain',),
 ('Eric Larson',),
 ('Paula McLain',),
 ('Steve Harvey',),
 ('Dinah Bucholz',),
 ('Ann Voskamp',),
 ('David McCullough',),
 ('Muriel Barbery',),
 ('American Psychiatric Association',),
 ('Mark Owen',),
 ('Todd Burpo',),
 ('Donna Tartt',),
 ('Adir Levy',),
 ('Margaret Atwood',),
 ('Suzanne Collins',),
 ('F. A. Hayek',),
 ('Stephenie Meyer',),
 ('Mike Moreno',),
 ('Ian K. Smith M.D.',),
 ('Alan Moore',),
 ('R. J. Palacio',),
 ('Tina Fey',),
 ("Madeleine L'Engle",),
 ('Cheryl Strayed',),
 ('John Grisham',),
 ('Peter A. Lillback',),
 ('Phil Ro

## VIEWing your query

You've worked hard to create the below SQL query:

    SELECT DISTINCT author AS unique_author
    FROM books;

What if you'd like to be able to refer to it later, or allow others to
access and use the results? The best way to do this is by creating a
view. Recall that a view is a virtual table: it's very similar to a real
table, but rather than the data itself being stored, the query code is
stored for later use.

**Instructions**

- Add a single line of code that saves the results of the written query
  as a view called `library_authors`.

**Answer**


In [8]:
%%sql
-- Save the results of this query as a view called library_authors
CREATE VIEW library_authors AS
SELECT DISTINCT author AS unique_author
FROM books;

-- Your code to create the view:
CREATE VIEW library_authors AS
SELECT DISTINCT author AS unique_author
FROM books;

-- Select all columns from library_authors
SELECT *
FROM library_authors;


 * postgresql://postgres:***@localhost/local
Done.
(psycopg2.errors.DuplicateTable) relation "library_authors" already exists

[SQL: -- Your code to create the view:
CREATE VIEW library_authors AS
SELECT DISTINCT author AS unique_author
FROM books;]
(Background on this error at: https://sqlalche.me/e/20/f405)


## Limiting results

Let's take a look at a few of the genres represented in our library's
books.

Recall that limiting results is useful when testing code since result
sets can have thousands of results! Queries are often written with a
`LIMIT` of just a few records to test out code before selecting
thousands of results from the database.

Let's practice with `LIMIT`!

**Instructions**

- Using PostgreSQL, select the `genre` field from the `books` table;
  limit the number of results to 10.

**Answer**


In [9]:
%%sql
-- Select the first 10 genres from books using PostgreSQL
SELECT genre
FROM books
LIMIT 10;


 * postgresql://postgres:***@localhost/local
10 rows affected.


[('Non Fiction',),
 ('Fiction',),
 ('Non Fiction',),
 ('Fiction',),
 ('Non Fiction',),
 ('Fiction',),
 ('Fiction',),
 ('Fiction',),
 ('Non Fiction',),
 ('Fiction',)]